telluric/georaster.py

```diff
@@ -747,22 +758,57 @@ def type_max(dtype):
         def type_min(dtype):
             return np.iinfo(dtype).min
 
-        if not np.issubdtype(dst_type, np.integer):
-            raise GeoRaster2Error('astype to non integer type is not supported - requested dtype: %s' % dst_type)
+        if (
+            in_range is None and out_range is not None or
+            out_range is None and in_range is not None
+        ):
+            raise GeoRaster2Error("Both ranges should be specified or none of them.")
 
         src_type = self.image.dtype
-        if not np.issubdtype(src_type, np.integer):
-            raise GeoRaster2Error('astype from non integer type is not supported - raster dtype: %s' % src_type)
-
-        if dst_type == src_type:
+        if not np.issubdtype(src_type, np.integer) and in_range == 'dtype':
+            in_range = 'image'
+            warnings.warn("Value 'dtype' of in_range parameter is supported only for integer type. "
+                          "Instead 'image' will be used.", GeoRaster2Warning)
+
+        if not np.issubdtype(dst_type, np.integer) and out_range == 'dtype':
+            raise GeoRaster2Error("Value 'dtype' of out_range parameter is supported only for integer type.")
+
+        if (
+            dst_type == src_type and
+            in_range == out_range == 'dtype'
+        ):
             return self
 
-        conversion_gain = \
-            (type_max(dst_type) - type_min(dst_type)) / (type_max(src_type) - type_min(src_type))
+        # streching or shrinking intensity levels is required
+        if out_range is not None:
+            if out_range == 'dtype':
+                omax = type_max(dst_type)
+                omin = type_min(dst_type)
+                if clip_negative and omin < 0:
+                    omin = 0
+            else:
+                omin, omax = out_range
+
+            if in_range == 'dtype':
+                imin = type_min(src_type)
+                imax = type_max(src_type)
+            elif in_range == 'image':
+                imin = min(self.min())
+                imax = max(self.max())
+            else:
+                imin, imax = in_range
+
+            if imin == imax:
+                conversion_gain = 0
+            else:
+                conversion_gain = (omax - omin) / (imax - imin)
+
+            # temp conversion, to handle saturation
+            dst_array = float(conversion_gain) * (self.image.astype(np.float) - imin) + omin
+            dst_array = np.clip(dst_array, omin, omax)
+        else:
+            dst_array = self.image
 
-        # temp conversion, to handle saturation
-        dst_array = conversion_gain * (self.image.astype(np.float) - type_min(src_type)) + type_min(dst_type)
-        dst_array = np.clip(dst_array, type_min(dst_type), type_max(dst_type))
         dst_array = dst_array.astype(dst_type)
         if stretch:
```

Comment:
I think it would be necessary to clarify the status of the "stretch histogram" functionality before we merge, in case we can decide what to do with it as part of this pull request.

Reply:
Does is mean that we can remove this part of the code from `astype`?

Reply:
I think so, we should be sure before that the behavior of `to_png` if not effected when running it on a raster with `nodata` pixels that effects the histogram, and of course you will need  to update the code of `to_png`, `to_bytes` and `_repr_png_` to use `in_range='image'` instead of `stretch=True`